pkg/builder/endpoints.go

```diff
@@ -0,0 +1,343 @@
+package builder
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	externaldns "sigs.k8s.io/external-dns/endpoint"
+
+	"github.com/kuadrant/dns-operator/api/v1alpha1"
+	"github.com/kuadrant/dns-operator/internal/common/hash"
+)
+
+const (
+	IPAddressType       AddressType = "IPAddress"
+	HostnameAddressType AddressType = "Hostname"
+	// HostnameRegex checks for at least two groups of allowed for URLs characters separated by "."
+	HostnameRegex = "^(?:[\\w\\-.~:\\/?#[\\]@!$&'()*+,;=]+)\\.(?:[\\w\\-.~:\\/?#[\\]@!$&'()*+,;=]+)$"
+
+	DefaultTTL      = 60
+	DefaultCnameTTL = 300
+
+	ClusterIDLength = 6
+
+	LabelLBAttributeGeoCode        = "kuadrant.io/lb-attribute-geo-code"
+	DefaultGeo              string = "default"
+	WildcardGeo             string = "*"
+)
+
+type AddressType string
+
+type TargetAddress struct {
+	Type  AddressType
+	Value string
+}
+
+// Target wraps a kubernetes ingress traffic resource e.g.Gateway, Ingress, Route etc.. but can wrap any resources
+// that has the desired geo and weight labels being applied, and can provide the required target addresses data.
+// This should be implemented as required for each type of ingress resource i.e. Gateway
+type Target interface {
+	metav1.Object
+	GetAddresses() []TargetAddress
+}
+
+type LoadBalancingSpec struct {
+	// DefaultWeight is the record weight to use when no other can be determined for a dns target cluster.
+	// The maximum value accepted is determined by the target dns provider.
+	DefaultWeight Weight `json:"defaultWeight"`
+
+	// list of CustomWeight selectors.
+	CustomWeights []*CustomWeight `json:"custom,omitempty"`
+
+	// defaultGeo is the country/continent/region code to use when no other can be determined for a dns target cluster.
+	// The values accepted are determined by the target dns provider, please refer to the appropriate docs below.
+	//
+	// Route53: https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resource-record-sets-values-geo.html
+	// Google: https://cloud.google.com/compute/docs/regions-zones
+	DefaultGeo string `json:"defaultGeo"`
+}
+
+type CustomWeight struct {
+	// Label selector to match resource storing custom weight attribute values e.g. kuadrant.io/lb-attribute-custom-weight: AWS.
+	Selector *metav1.LabelSelector `json:"selector"`
+
+	// The weight value to apply when the selector matches.
+	Weight Weight `json:"weight"`
+}
+
+// Weight can only be a zero or positive integer
+// +kubebuilder:validation:Minimum=0
+type Weight int
```

Comment:
I was trying to avoid having them in the `v1alpha1` because it has nothing to do with DNSRecord definition and is used only in the endpoints generation :thinking:  and in `kuadrant-operator` we would import builder package and `v1alpa1` anyway. 
My reasoning was that by keeping it in the builder package we would be able to change it separately from API changes :thinking:

Reply:
I'm not saying its needed for DNSRecord or that it should be in the dnsrecord_types.

Presumably the DNSPolicy will reference the LoadBalancingSpec from this repo The kuadrant operator will eventually end up with multiple versions of the DNSPolicy(v1alpha1, v1alpha2, v1 etc..), all of these could exist at the same time inside the same version of kuadrant operator with each converting to the latest.  If we need to make changes to LoadBalancingSpec for one version but not change the rest of them, how do you see that working?

Reply:
I see, moved. Thank you for explanation